scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala

```diff
@@ -398,10 +457,24 @@ object TreeSyntax {
       case t: Term.Tuple => m(SimpleExpr1, s("(", r(t.args, ", "), ")"))
       case t: Term.Block =>
         import Term.{Block, Function}
-        def block(pre: Show.Result, x: Seq[Stat]) = {
+        def requireNewLines(stats: List[Stat]): List[Show.Result] =
+          stats
+            .foldLeft((false, List.empty[Show.Result])) {
+              case ((req, acc), block: Block) => (false, r(if (req) EOL else "", i(block)) :: acc)
+              case ((_, acc), stat) => (guessNeedsLineSep(stat), i(stat) :: acc)
+            }
+            ._2
+            .reverse
+        def block(pre: Show.Result, x: List[Stat]) = {
           val res =
             if (pre == Show.None && x.isEmpty) s("{}")
-            else s("{", w(" ", pre), r(x.map(i(_))), n("}"))
+            else
+              s(
+                "{",
+                w(" ", pre),
+                r(if (dialect.allowSignificantIndentation) x.map(i(_)) else requireNewLines(x)),
+                n("}")
+              )
           m(SimpleExpr, res)
         }
```

Comment:
I think this is simpler and checks only on demand:
```
def withBlankLines(x: Seq[Stat]) = {
  val builder = List.newBuilder[Show.Result]
  var prevStat = null
  x.foreach { stat =>
    if (stat.is[Block] && null != prevStat && guessNeedsLineSep(prevStat)) builder += EOL
    builder += i(stat)
    prevStat = stat
  }
  builder.result()
}

def block(pre: Show.Result, x: Seq[Stat]) = {
  val res =
    if (pre == Show.None && x.isEmpty) s("{}")
    else {
      val stats = if (dialect.allowSignificantIndentation) x.map(i) else withBlankLines(x)
      s("{", w(" ", pre), r(stats), n("}"))
    }
  m(SimpleExpr, res)
}
```

Reply:
Checking only on demand is definitely better, as this Block cases are rare. The list builder is also better than my concatenation + reverse thing I had.